include/deal.II/numerics/smoothness_estimator.templates.h

```diff
@@ -0,0 +1,473 @@
+      dof_handler.get_fe_collection().max_degree();
+
+    Table<dim, number_coeff> expansion_coefficients;
+    resize(expansion_coefficients, max_degree);
+
+    Vector<number>                                            local_dof_values;
+    std::vector<double>                                       ln_k;
+    std::pair<std::vector<unsigned int>, std::vector<double>> res;
+    for (auto &cell : dof_handler.active_cell_iterators())
+      if (cell->is_locally_owned())
+        {
+          local_dof_values.reinit(cell->get_fe().dofs_per_cell);
+
+          auto solution_it              = all_solutions.cbegin();
+          auto smoothness_indicators_it = all_smoothness_indicators.begin();
+          for (; solution_it != all_solutions.cend();
+               ++solution_it, ++smoothness_indicators_it)
+            {
+              // Inside the loop, we first need to get the values of the local
+              // degrees of freedom and then need to compute the series
+              // expansion by multiplying this vector with the matrix ${\cal F}$
+              // corresponding to this finite element.
+              cell->get_dof_values(*(*solution_it), local_dof_values);
+
+              fe_series.calculate(local_dof_values,
+                                  cell->active_fe_index(),
+                                  expansion_coefficients);
+
+              // We fit our exponential decay of expansion coefficients to the
+              // provided regression_strategy on each possible value of |k|. To
+              // this end, we use FESeries::process_coefficients() to rework
+              // coefficients into the desired format.
+              res = FESeries::process_coefficients<dim>(
+                expansion_coefficients,
+                std::bind(&predicate_k_in_r<dim>,
+                          std::placeholders::_1,
+                          max_degree),
+                regression_strategy);
+
+              Assert(res.first.size() == res.second.size(), ExcInternalError());
+
+              // Prepare linear equation for the logarithmic least squares fit.
+              //
+              // First, calculate ln(|k|). This vector will be the same for all
+              // the cells so we can calculate ln(|k|) only once.
+              //
+              // For Fourier expansion, this translates to
+              // ln(2*pi*sqrt(predicate)) = ln(2*pi) + 0.5*ln(predicate). Since
+              // we are just interested in the slope of a linear regression
+              // later, we omit the ln(2*pi) factor.
+              if (ln_k.empty())
+                {
+                  ln_k.resize(res.first.size());
+                  for (unsigned int f = 0; f < res.first.size(); ++f)
+                    ln_k[f] = 0.5 * std::log((double)res.first[f]);
+                }
+
+              // Second, calculate ln(U_k).
+              for (auto &residual_element : res.second)
+                residual_element = std::log(residual_element);
```

Comment:
While preparing a unit test for the Fourier smoothness estimation, I ran into a problem in this line: Whenever the `residual_element` is zero, the logarithm fails. I tried to fix it with a conditional
```
for (auto &residual_element : res.second)
  if (residual_element > 0)
    residual_element = std::log(residual_element);
  else
    residual_element = std::log(std::numeric_limits<double>::min());
```
Using `-1. * std::numeric_limits<double>::min()` would fail later when `FESeries::linear_regression()` sums up the residuals.

Further, this issue is runtime-dependent, i.e. the residual is not always zero, but close to it.

Do you think this is a valid fix? Alternatively, I could check up front if any residual is zero. If that is the case, I assume the function to be in C^infty, and return `std::numeric_limits<float>::max()`. What do you think about that approach?

Reply:
> I ran into a problem in this line: Whenever the `residual_element` is zero, the logarithm fails

yes, that's expected. I think the logic with logarithms should be changed here so that you altogether skip points for which `log(small)` happens. At least that's what I opted for `Legendre` algorithm, see the lines above 

```
FESeries::linear_regression(x, y)
```

Reply:
I think you should test whether the residual value is small not only in absolute terms, but also relative to, say, the largest element of the bunch.

Reply:
This piece of code does not seem to be in the patch any more. Can you point me to where you have it now?

In general, can you say what happens when "the logarithm fails"?

Reply:
That section of the code moved here:
https://github.com/dealii/dealii/blob/3a23f19f109972a850794c093eba478965ae466b/source/numerics/smoothness_estimator.cc#L345-L347

The expansion coefficients will be too small at some stage, i.e. too close to zero, which will cause `std::log` to fail. Thus, @davydden and me only consider those coefficients larger than an absolute value for the linear regression, and skip them otherwise. You proposed to check for a relative value in addition.

I am not entirely sure if we should extend this PR by this relative value check, or leave it for a later stage for now.

Reply:
If you're happy with it as is, let's stick with it until we have a good reason to change anything.

Reply:
But in general, *nothing* should use absolute tolerances. That's because if you think of a solution `u`, then you would expect all refinement decisions to be exactly the same if you had `10^12 u` or `10^{-12} u`. Smoothness is a quality of a solution that is independent of its magnitude.